src/compiler/checker.ts

```diff
@@ -6510,14 +6502,21 @@ namespace ts {
                 const typePredicate = declaration.type && declaration.type.kind === SyntaxKind.TypePredicate ?
                     createTypePredicateFromTypePredicateNode(declaration.type as TypePredicateNode) :
                     undefined;
-                // JS functions get a free rest parameter if they reference `arguments`
+                // JS functions get a free rest parameter if:
+                // a) The last parameter has `...` preceding its type
+                // b) It references `arguments` somewhere
                 let hasRestLikeParameter = hasRestParameter(declaration);
-                if (!hasRestLikeParameter && isInJavaScriptFile(declaration) && containsArgumentsReference(declaration)) {
-                    hasRestLikeParameter = true;
-                    const syntheticArgsSymbol = createSymbol(SymbolFlags.Variable, "args" as __String);
-                    syntheticArgsSymbol.type = anyArrayType;
-                    syntheticArgsSymbol.isRestParameter = true;
-                    parameters.push(syntheticArgsSymbol);
+                if (!hasRestLikeParameter && isInJavaScriptFile(declaration)) {
+                    const lastParam = lastOrUndefined(declaration.parameters);
+                    const lastParamIsDocumentedAsRest = !!lastParam && some(getJSDocParameterTags(lastParam), p =>
+                        p.typeExpression && p.typeExpression.type.kind === SyntaxKind.JSDocVariadicType);
+                    if (lastParamIsDocumentedAsRest || containsArgumentsReference(declaration)) {
+                        hasRestLikeParameter = true;
+                        const syntheticArgsSymbol = createSymbol(SymbolFlags.Variable, "args" as __String);
+                        syntheticArgsSymbol.type = lastParamIsDocumentedAsRest ? createArrayType(getTypeOfParameter(lastParam.symbol)) : anyArrayType;
+                        syntheticArgsSymbol.isRestParameter = true;
+                        parameters.push(syntheticArgsSymbol);

```

Comment:
This will make f(a) have the signature f(a, ...Args)when the intent is just to have f(...Args). That is, the intent of jsdoc `{...number} a` is to replace a with ...args.

Reply:
After thinking about it, the naive fix of replacing the actual symbol with the synthetic one is likely to cause problems with, eg, renaming a. So this code might be good enough. Do you mind checking how the naive fix behaves?

Reply:
The real problem is, what will happen if the user tries to access `a` inside the body? Should we be giving them `number | undefined` because they declared it as a rest parameter?